src/engine_constants/mod.rs

```diff
@@ -1575,31 +1575,42 @@ impl EngineConstants {
-                    (Axis::TriggerRight, GamepadConsts::rects(Rect::new(32, 80, 64, 96))),
-                ]),
+            gamepad: {
+                let mut holder = GamepadConsts {
+                    button_rects: HashMap::from([
+                        (Button::North, GamepadConsts::rects(Rect::new(0, 0, 32, 16))),
+                        (Button::South, GamepadConsts::rects(Rect::new(0, 16, 32, 32))),
+                        (Button::East, GamepadConsts::rects(Rect::new(0, 32, 32, 48))),
+                        (Button::West, GamepadConsts::rects(Rect::new(0, 48, 32, 64))),
+                        (Button::DPadDown, GamepadConsts::rects(Rect::new(0, 64, 32, 80))),
+                        (Button::DPadUp, GamepadConsts::rects(Rect::new(0, 80, 32, 96))),
+                        (Button::DPadRight, GamepadConsts::rects(Rect::new(0, 96, 32, 112))),
+                        (Button::DPadLeft, GamepadConsts::rects(Rect::new(0, 112, 32, 128))),
+                        (Button::LeftShoulder, GamepadConsts::rects(Rect::new(32, 32, 64, 48))),
+                        (Button::RightShoulder, GamepadConsts::rects(Rect::new(32, 48, 64, 64))),
+                        (Button::Start, GamepadConsts::rects(Rect::new(32, 96, 64, 112))),
+                        (Button::Back, GamepadConsts::rects(Rect::new(32, 112, 64, 128))),
+                        (Button::LeftStick, GamepadConsts::rects(Rect::new(32, 0, 64, 16))),
+                        (Button::RightStick, GamepadConsts::rects(Rect::new(32, 16, 64, 32))),
+                    ]),
+                    axis_rects: HashMap::from([
+                        (Axis::LeftX, GamepadConsts::rects(Rect::new(32, 0, 64, 16))),
+                        (Axis::LeftY, GamepadConsts::rects(Rect::new(32, 0, 64, 16))),
+                        (Axis::RightX, GamepadConsts::rects(Rect::new(32, 16, 64, 32))),
+                        (Axis::RightY, GamepadConsts::rects(Rect::new(32, 16, 64, 32))),
+                        (Axis::TriggerLeft, GamepadConsts::rects(Rect::new(32, 64, 64, 80))),
+                        (Axis::TriggerRight, GamepadConsts::rects(Rect::new(32, 80, 64, 96))),
+                    ]),
+                };
+                // Swap x-y and a-b buttons on nintendo map (must be done here to comply with the nicalis button table)
+                let button = holder.button_rects.get(&Button::North).unwrap()[3].clone();
+                holder.button_rects.get_mut(&Button::North).unwrap()[3] = holder.button_rects.get(&Button::West).unwrap()[3];
+                holder.button_rects.get_mut(&Button::West).unwrap()[3] = button;
+                let button = holder.button_rects.get(&Button::South).unwrap()[3].clone();
+                holder.button_rects.get_mut(&Button::South).unwrap()[3] = holder.button_rects.get(&Button::East).unwrap()[3];
+                holder.button_rects.get_mut(&Button::East).unwrap()[3] = button;
```

Comment:
This makes like no sense at all? There's no condition for Nintendo controllers and this has the same effect as swapping rectangles in previous code.

Reply:
[Nintendo controller mapping comes from rect index 3](https://github.com/doukutsu-rs/doukutsu-rs/blob/c19d0738fa964ae0df382dffe14546c0c1250797/src/framework/gamepad.rs#L201), which is the index of the buttons I'm swapping. The other indexes, 0 through 2 are left unmodified. They would be changed as well if I simply swapped definitions in the old code.

The current code creates a set of 4 `Rect`s for each `Button` in the hashmap. (see [`GampadConsts::rects`](https://github.com/doukutsu-rs/doukutsu-rs/blob/c19d0738fa964ae0df382dffe14546c0c1250797/src/engine_constants/mod.rs#L237)) This code assumes the buttons for each system will be in the same order, but if this were the case, the buttons for the [Nintendo controllers would be reversed](https://github.com/doukutsu-rs/doukutsu-rs/blob/master/src/data/builtin/builtin_data/buttons.png). Swapping the buttons in the image file isn't an option because d-rs loads the button image assets from the Nicalis ports, which can't be distributed with the engine. The game has to bend to accommodate those assets.